### PR TITLE
Feature/get most relevant jira ticket

### PR DESCRIPTION
**Pull request by baristaGeek**
## Description
To reinforce our core concepts of code archeology and going beyond git blame to help devs understand code context, we're building this new feature. Get the most relevant jira ticket for a given line, block or file of code. 

The steps are:
1. Get the commit hashes and most relevant PR title with what we were already doing before
2. Send the PR title (with stop words removed) as a search query to our API
3. Render the Jira ticket that has the most word matches, regardless of its ticket state. 

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)  
- [x] New feature (non-breaking change which adds functionality)  
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)  
- [ ] Documentation update  
- [ ] Chore: cleanup/renaming, etc  
- [ ] RFC  

## Notes
- Still missing a programmatically retrieved user email
- Still missing how to address the case where the user is authed but is not an actively paying Watermelon user

### CODE REVIEW

**Comment by baristaGeek** [comment]
@EstebanDalelR I guess https://github.com/watermelontools/watermelon/pull/82/files is the answer to the second question I left on the Notes section?

File: media/main.js [hunk 1]
```diff
@@ -56,10 +58,14 @@ function handleMessage(message) {
       if (message.owner && message.repo) {
         commitLink = `https://github.com/${message.owner}/${message.repo}/commit/`;
       }
+      // blame table
       addBlametoDoc(message.data.uniqueBlames, commitLink);
       // prs
       webviewDebugLogger(`Received prs: ${JSON.stringify(message.data)}`);
       addPRsToDoc(message.data.sortedPRs);
+      // jira
+      console.log("most relevant jira ticket - main.js: ", message.data.mostRelevantJiraTicket);
```
Comment (EstebanDalelR) [inline media/main.js:67]: Please use the debug logger.

File: package.json [hunk 2]
```diff
@@ -298,7 +305,8 @@
     "@vscode/test-electron": "^2.1.2",
     "axios": "^0.26.1",
     "git-url-parse": "^11.6.0",
-    "octokit": "^1.7.1"
+    "octokit": "^1.7.1",
+    "uuid": "^8.3.2"
```
Comment (EstebanDalelR) [inline package.json:309]: Why uuid?

File: src/utils/jira/getMostRelevantJiraTicket.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import axios from "axios";
+import { backendURL } from "../../constants";
+
+export default async function getMostRelevantJiraTicket({
+  userEmail,
+  prTitle,
+}: {
+  userEmail: string;
+  prTitle: string;
```
Comment (EstebanDalelR) [inline src/utils/jira/getMostRelevantJiraTicket.ts:9]: This depends on github?

Comment (EstebanDalelR) [reply src/utils/jira/getMostRelevantJiraTicket.ts:9]: What if the user does not have github?

**Review by EstebanDalelR — CHANGES_REQUESTED** [review]
Few questions open

File: src/utils/jira/getMostRelevantJiraTicket.ts (hunk 3 above)
Comment (baristaGeek) [reply src/utils/jira/getMostRelevantJiraTicket.ts:9]: If the user doesn't have GitHub, he wouldn't be able to get to try our product for free in the first place. So yes, this depends on GitHub. 

As we expand to the counterparts (Bitbucket, Notion, etc.) we can build some sort of middle layer that allows to connect any git system with any ticketing system. For the moment, let's go deep on the VS Code + GitHub + Jira integration.

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,8 +4,11 @@ All notable changes to the "watermelon" extension will be documented in this fil
 
 Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.
 
-## [1.6.0]
+## [1.6.1]
+- Get code context button now retrieves the most relevant Jira ticket, if the user is authenticated with Jira
+- Watermelon is now an authentication provider, to allow us to connect to all your services
```
Comment (EstebanDalelR) [inline CHANGELOG.md:9]: Why is this declared again?

File: media/utils/addMostRelevantJiraTicket.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+const addMostRelevantJiraTicket = (jiraTicket) => {
+  $("#mostRelevantJiraTicketHolder").append(`<h3>Most Relevant Jira Ticket</h3>`);
+  $("#mostRelevantJiraTicketHolder").append(`
```
Comment (EstebanDalelR) [inline media/utils/addMostRelevantJiraTicket.js:3]: What happens if it's empty?

File: media/main.js [hunk 6]
```diff
@@ -56,10 +58,13 @@ function handleMessage(message) {
       if (message.owner && message.repo) {
         commitLink = `https://github.com/${message.owner}/${message.repo}/commit/`;
       }
+      // blame table
       addBlametoDoc(message.data.uniqueBlames, commitLink);
       // prs
       webviewDebugLogger(`Received prs: ${JSON.stringify(message.data)}`);
       addPRsToDoc(message.data.sortedPRs);
+      // jira
+      addMostRelevantJiraTicket(message.data.mostRelevantJiraTicket);
```
Comment (EstebanDalelR) [inline media/main.js:67]: What happens if it's empty?

File: package.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "watermelon-tools",
   "description": "Watermelon, find all the historical context of your code",
-  "version": "1.6.0",
+  "version": "1.6.1",
```
Comment (EstebanDalelR) [inline package.json:4]: Why is a PR changing the version if it's not from `dev` to `main`?

File: package.json [hunk 8]
```diff
@@ -306,8 +306,7 @@
     "@vscode/test-electron": "^2.1.2",
     "axios": "^0.26.1",
     "git-url-parse": "^11.6.0",
-    "octokit": "^1.7.1",
-    "uuid": "^8.3.2"
```
Comment (EstebanDalelR) [inline package.json:310]: thank you!

File: src/extension.ts [hunk 9]
```diff
@@ -205,9 +205,15 @@ export async function activate(context: vscode.ExtensionContext) {
         );
         let uniqueBlames = await getBlame(gitAPI, startLine, endLine);
 
+        // Jira
+        const mostRelevantJiraTicket = await getMostRelevantJiraTicket({
+          userEmail: session.account.label,
+          prTitle: sortedPRs[0].title,
```
Comment (EstebanDalelR) [inline src/extension.ts:211]: If there is no PR title, should we still call it?

File: src/utils/jira/getMostRelevantJiraTicket.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+}) {
+  const jiraTickets = await axios.post(
+    `${backendURL}/api/jira/getMostRelevantJiraTicket`,
+    {
+      userEmail,
+      prTitle,
+    }
+  ).then(res => res.data);
+
+  return jiraTickets[0];
```
Comment (EstebanDalelR) [inline src/utils/jira/getMostRelevantJiraTicket.ts:19]: There is no "see more"?

File: media/utils/addMostRelevantJiraTicket.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    <div class="Box-header">
+      <h5 class="Box-title Truncate">
+        ${jiraTicket?.key}
+      </h5>
+      <span style="color: green">${jiraTicket?.fields?.status?.name}</span>
+      <p>${jiraTicket?.fields?.summary}</p>
+    </div>
+  `);
+};
+
```
Comment (EstebanDalelR) [inline media/utils/addMostRelevantJiraTicket.js:14]: Is there no "see more"?

**Review by EstebanDalelR — CHANGES_REQUESTED** [review]

File: src/extension.ts (hunk 9 above)
Comment (baristaGeek) [reply src/extension.ts:211]: This is a very good question. In the new commit I propose the following (would like your input): 

- Send a pr title, wrangle it on the back-end, and search from there
- If there is no PR Title, send an empty String
- In the latter case, send an empty string, which will return no ticket. In this case no Jira ticket will be retrieved. 
- Then on main.js line 66, The control flow doesn't enter the if (message.data.mostRelevantJiraTicket) condition and addMostRelevantJiraTicket( ) isn't called.

Comment (EstebanDalelR) [reply src/extension.ts]: Ok, so no fallback to commit message?

Comment (baristaGeek) [reply src/extension.ts]: Is it possible to have a commit message, withou having a PR Title?

File: CHANGELOG.md [hunk 12]
```diff
@@ -5,11 +5,11 @@ All notable changes to the "watermelon" extension will be documented in this fil
 Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.
 
 ## [1.6.1]
-
+- Get code context button now retrieves the most relevant Jira ticket, if the user is authenticated with Jira
+- Watermelon is now an authentication provider, to allow us to connect to all your services
```
Comment (EstebanDalelR) [inline CHANGELOG.md:9]: Why is this repeated?

File: media/main.js [hunk 13]
```diff
@@ -56,10 +58,15 @@ function handleMessage(message) {
       if (message.owner && message.repo) {
         commitLink = `https://github.com/${message.owner}/${message.repo}/commit/`;
       }
+      // blame table
       addBlametoDoc(message.data.uniqueBlames, commitLink);
       // prs
       webviewDebugLogger(`Received prs: ${JSON.stringify(message.data)}`);
       addPRsToDoc(message.data.sortedPRs);
+      // jira
+      if (message.data.mostRelevantJiraTicket) {
```
Comment (EstebanDalelR) [inline media/main.js:67]:
```suggestion
if (message.data?.mostRelevantJiraTicket) {
```

File: media/utils/addMostRelevantJiraTicket.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+const addMostRelevantJiraTicket = (jiraTicket) => {
+  $("#mostRelevantJiraTicketHolder").append(`<h3>Most Relevant Jira Ticket</h3>`);
+  $("#mostRelevantJiraTicketHolder").append(`
+  <h3>Assigned Jira Tickets</h3>
```
Comment (EstebanDalelR) [inline media/utils/addMostRelevantJiraTicket.js:4]: Why this other title? Do I only get the ones assigned to me?

File: src/extension.ts [hunk 15]
```diff
@@ -205,10 +205,26 @@ export async function activate(context: vscode.ExtensionContext) {
         );
         let uniqueBlames = await getBlame(gitAPI, startLine, endLine);
 
-        provider.sendMessage({
-          command: "prs",
-          data: { sortedPRs, uniqueBlames },
-        });
+        const parsedCommitMessage = new Object(uniqueBlames[0]) as any;
```
Comment (EstebanDalelR) [inline src/extension.ts:208]: Why not set it as Object? Or better yet, shape it?

File: src/extension.ts [hunk 16]
```diff
@@ -205,10 +205,26 @@ export async function activate(context: vscode.ExtensionContext) {
...
+
+        // Jira
+        const mostRelevantJiraTicket = await getMostRelevantJiraTicket({
+          userEmail: session.account.label,
+          prTitle: sortedPRs[0].title || parsedCommitMessage,
```
Comment (EstebanDalelR) [inline src/extension.ts:213]: Why send the whole object, not the commit title?

**Review by EstebanDalelR — CHANGES_REQUESTED** [review]

**Review by EstebanDalelR — APPROVED** [review]

**Pull request merged**